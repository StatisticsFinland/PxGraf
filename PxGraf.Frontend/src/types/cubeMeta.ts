/* istanbul ignore file */

import { MultiLanguageString } from "./multiLanguageString";

/**
 * Interface for cube meta properties. A cube is a multi-dimensional result of Px file data filtered by a query.
 * @property {string[]} languages - List of available languages.
 * @property {MultiLanguageString} header - The multi language header of the cube.
 * @property {MultiLanguageString} note - The note of the cube.
 * @property {IVariable[]} variables - List of variables.
 */
export interface ICubeMeta {
    languages: string[],
    header: MultiLanguageString,
    note: MultiLanguageString,
    variables: IVariable[]
}

/**
 * Interface for variable properties. Each variable defines a dimension in a cube.
 * @property {string} code - The code of the variable.
 * @property {MultiLanguageString} name - The multi language name of the variable.
 * @property {MultiLanguageString} note - The multi language note of the variable.
 * @property {VariableType} type - The type of the variable.
 * @property {IVariableValue[]} values - List of values associated with this variable.
 */
export interface IVariable {
    code: string,
    name: MultiLanguageString,
    note: MultiLanguageString,
    type: VariableType,
    values: IVariableValue[]
}

/**
 * Interface for variable value properties.
 * @property {string} code - The code of the value.
 * @property {MultiLanguageString} name - The multi language name of the value.
 * @property {MultiLanguageString} note - The multi language note of the value.
 * @property {boolean} isSum - Flag to indicate if the value is an elimination value.
 * @property {IContentComponent} contentComponent - The content variable component of the value.
 * @property {IVirtualComponent[]} virtualComponents - List of virtual components.
 */
export interface IVariableValue {
    code: string,
    name: MultiLanguageString,
    note: MultiLanguageString,
    isSum: boolean,
    contentComponent?: IContentComponent
    virtualComponents?: IVirtualComponent[]
}

/**
 * Interface for value content variable value properties.
 * @property {MultiLanguageString} unit - The unit of the value.
 * @property {MultiLanguageString} source - The source of the value.
 * @property {number} numberOfDecimals - The precision, number of decimals of the value.
 * @property {string} lastUpdated - The last updated date of the value.
 */
export interface IContentComponent {
    unit: MultiLanguageString,
    source: MultiLanguageString,
    numberOfDecimals: number,
    lastUpdated: string
}

/**
 * Interface for virtual component properties.
 * @property {string} operator - The operator of the virtual component.
 * @property {string[]} operandCoded - List of coded operands.
 */
export interface IVirtualComponent {
    operator: string,
    operandCoded: string[]
}

/**
 * Enumeration for variable types.
 */
export enum VariableType {
    Unknown = 'N',
    Time = 'T',
    Ordinal = 'P',
    Geological = 'G',
    Content = 'C',
    OtherClassificatory = 'F',
}