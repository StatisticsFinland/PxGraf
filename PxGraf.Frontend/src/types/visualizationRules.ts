/* istanbul ignore file */

import { MultiLanguageString } from "./multiLanguageString"

/**
 * Interface for visualization rules.
 * @property allowManualPivot - Flag to indicate if manual pivot is allowed.
 * @property sortingOptions - List of sorting options.
 * @property multiselectVariableAllowed - Flag to indicate if multiselect variable is allowed.
 * @property visualizationTypeSpecificRules - Object for visualization type specific rules.
 */
export interface IVisualizationRules {
    allowManualPivot: boolean,
    sortingOptions: ISortingOption[],
    multiselectVariableAllowed: boolean
    visualizationTypeSpecificRules?: ITypeSpecificVisualizationRules
}

/**
 * Interface for different sorting options
 * @property code - Code for the sorting option
 * @property description - Description for the sorting option as a multi language string
 */
export interface ISortingOption {
    code: string,
    description: MultiLanguageString
}

/**
 * Interface for visualization type specific rules.
 * @property allowShowingDataPoints - Flag to indicate if showing data points is allowed.
 * @property allowCuttingYAxis - Flag to indicate if cutting the Y-axis is allowed.
 * @property allowMatchXLabelsToEnd - Flag to indicate if matching X labels to the end is allowed.
 * @property allowSetMarkerScale - Flag to indicate if setting marker scale is allowed.
 */
export interface ITypeSpecificVisualizationRules {
    allowShowingDataPoints: boolean,
    allowCuttingYAxis: boolean,
    allowMatchXLabelsToEnd: boolean,
    allowSetMarkerScale: boolean
}