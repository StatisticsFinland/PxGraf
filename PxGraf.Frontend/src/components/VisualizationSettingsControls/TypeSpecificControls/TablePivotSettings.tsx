import { Stack, IconButton, FormControl } from '@mui/material';

import React, { useState } from 'react';

import ArrowDownwardIcon from '@mui/icons-material/ArrowDownward';
import ArrowUpwardIcon from '@mui/icons-material/ArrowUpward';
import SwapHorizIcon from '@mui/icons-material/SwapHoriz';

import { useTranslation } from 'react-i18next';
import VariableList from '../UtilityComponents/VariableList';
import { IVisualizationSettingsProps } from '../VisualizationSettingsControl';
import { IDimension } from 'types/cubeMeta';
import styled from 'styled-components';
import { Query } from "types/query";

const VariableListWrapper = styled(Stack)`
    padding: 8px;
    align-items: center;
    min-width: 375px;
`;

const IconWrapper = styled(Stack)`
    min-width: 75px;
`;

interface ITableSettingsProps extends IVisualizationSettingsProps {
    variables: IDimension[]
    selectableVariables?: IDimension[]
    query?: Query
}

export const TablePivotSettings: React.FC<ITableSettingsProps> = ({ visualizationSettings, settingsChangedHandler, variables, selectableVariables, query }) => {
    const { t } = useTranslation();
    const [selected, setSelected] = useState("");

    const upHandler = () => {
        if (selected == null) return;
        if (visualizationSettings.columnVariableCodes.includes(selected) && visualizationSettings.columnVariableCodes[0] !== selected) {
            const varsCopy = [...visualizationSettings.columnVariableCodes];
            const i = varsCopy.indexOf(selected);
            varsCopy[i] = varsCopy[i - 1];
            varsCopy[i - 1] = selected;
            settingsChangedHandler({ ...visualizationSettings, columnVariableCodes: varsCopy });
        }
        else if (visualizationSettings.rowVariableCodes.includes(selected) && visualizationSettings.rowVariableCodes[0] !== selected) {
            const varsCopy = [...visualizationSettings.rowVariableCodes];
            const i = varsCopy.indexOf(selected);
            varsCopy[i] = varsCopy[i - 1];
            varsCopy[i - 1] = selected;
            settingsChangedHandler({ ...visualizationSettings, rowVariableCodes: varsCopy });
        }
    }

    const downHandler = () => {
        if (selected == null) return;
        if (visualizationSettings.columnVariableCodes.includes(selected) && visualizationSettings.columnVariableCodes[visualizationSettings.columnVariableCodes.length - 1] !== selected) {
            const varsCopy = [...visualizationSettings.columnVariableCodes];
            const i = varsCopy.indexOf(selected);
            varsCopy[i] = varsCopy[i + 1];
            varsCopy[i + 1] = selected;
            settingsChangedHandler({ ...visualizationSettings, columnVariableCodes: varsCopy });
        }
        else if (visualizationSettings.rowVariableCodes.includes(selected) && visualizationSettings.rowVariableCodes[visualizationSettings.rowVariableCodes.length - 1] !== selected) {
            const varsCopy = [...visualizationSettings.rowVariableCodes];
            const i = varsCopy.indexOf(selected);
            varsCopy[i] = varsCopy[i + 1];
            varsCopy[i + 1] = selected;
            settingsChangedHandler({ ...visualizationSettings, rowVariableCodes: varsCopy });
        }
    }

    const swapHandler = () => {
        const rowCopy = [...visualizationSettings.rowVariableCodes];
        const columnCopy = [...visualizationSettings.columnVariableCodes];
        const rowIndex = rowCopy.indexOf(selected);
        const columnIndex = columnCopy.indexOf(selected);
        if (rowIndex > -1) {
            rowCopy.splice(rowIndex, 1);
            settingsChangedHandler({ ...visualizationSettings, rowVariableCodes: rowCopy, columnVariableCodes: [...columnCopy, selected] });
        } else if (columnIndex > -1) {
            columnCopy.splice(columnIndex, 1);
            settingsChangedHandler({ ...visualizationSettings, rowVariableCodes: [...rowCopy, selected], columnVariableCodes: columnCopy });
        }
    }

    const getMultiValueVariables = (codes): IDimension[] => {
        //Filter multi value variables and variables that are defined with either "from" or "all" value filters. Exclude selectables - except for multivalueseletable variable
        const multiValueVariableCodes = codes
            .map(c => variables.find(v => v.Code === c))
            .filter(v => variables.filter(v => query[v.Code].valueFilter.type === 'from' || query[v.Code].valueFilter.type === 'all')?.includes(v) || v.Values.length > 1)
            .filter(v => !selectableVariables?.includes(v) || v.Code === visualizationSettings.multiselectableVariableCode);
        return multiValueVariableCodes;
    }

    return (
        <FormControl fullWidth>
            <VariableListWrapper direction="row">
                <VariableList
                    title={t("chartSettings.rowVariables")}
                    variables={getMultiValueVariables(visualizationSettings.rowVariableCodes)}
                    selectedVariableCode={selected}
                    selectedChangedHandler={(newSel) => setSelected(newSel)}
                />
                <IconWrapper>
                    <IconButton aria-label={t('tableSettings.up')} onClick={() => upHandler()}>
                        <ArrowUpwardIcon />
                    </IconButton>
                    <IconButton aria-label={t('tableSettings.down')} onClick={() => downHandler()}>
                        <ArrowDownwardIcon />
                    </IconButton>
                    <IconButton aria-label={t('tableSettings.swap')} onClick={() => swapHandler()}>
                        <SwapHorizIcon />
                    </IconButton>
                </IconWrapper>
                <VariableList
                    title={t("chartSettings.columnVariables")}
                    variables={getMultiValueVariables(visualizationSettings.columnVariableCodes)}
                    selectedVariableCode={selected}
                    selectedChangedHandler={(newSel) => setSelected(newSel)}
                />
            </VariableListWrapper>
        </FormControl>
    );
}

export default TablePivotSettings;