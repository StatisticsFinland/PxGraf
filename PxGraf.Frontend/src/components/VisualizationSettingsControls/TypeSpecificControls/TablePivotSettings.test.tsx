import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { TablePivotSettings } from "./TablePivotSettings";
import { IDimension, EDimensionType } from "types/cubeMeta";
import UiLanguageContext from 'contexts/uiLanguageContext';
import { Query, FilterType } from "types/query";
import '@testing-library/jest-dom';
import { IVisualizationRules } from '../../../types/visualizationRules';
import { IVisualizationSettings } from '../../../types/visualizationSettings';

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => null),
            },
        };
    },
}));

const mockVisualizationRules: IVisualizationRules = {
    allowManualPivot: false,
    sortingOptions: null,
    multiselectDimensionAllowed: false
};

const mockVisualizationSettings: IVisualizationSettings = {
    defaultSelectableVariableCodes: null,
    pivotRequested: null,
    cutYAxis: null,
    multiselectableVariableCode: null,
    rowVariableCodes: ["foobar1", "foobar3"],
    columnVariableCodes: ["foobar2", "foobar4"],
    sorting: null,
    matchXLabelsToEnd: false,
    markerSize: null,
}

const mockDimensions: IDimension[] = [
    {
        code: "foobar1",
        name: { fi: "foo1", sv: "bar1", en: "foobar1" },
        type: EDimensionType.Other,
        values: [
            {
                code: "barfoo1",
                name: { fi: "fyy1", sv: "bör1", en: "fyybör1" },
                isVirtual: false
            },
            {
                code: "barfoo2",
                name: { fi: "fyy2", sv: "bör2", en: "fyybör2" },
                isVirtual: false
            }
        ]
    },
    {
        code: "foobar2",
        name: { fi: "foo2", sv: "bar2", en: "foobar2" },
        type: EDimensionType.Other,
        values: [
            {
                code: "barfoo1",
                name: { fi: "fyy1", sv: "bör1", en: "fyybör1" },
                isVirtual: false
            },
            {
                code: "barfoo2",
                name: { fi: "fyy2", sv: "bör2", en: "fyybör2" },
                isVirtual: false
            }
        ]
    },
    {
        code: "foobar3",
        name: { fi: "foo3", sv: "bar3", en: "foobar3" },
        type: EDimensionType.Other,
        values: [
            {
                code: "barfoo1",
                name: { fi: "fyy1", sv: "bör1", en: "fyybör1" },
                isVirtual: false
            },
            {
                code: "barfoo2",
                name: { fi: "fyy2", sv: "bör2", en: "fyybör2" },
                isVirtual: false
            }
        ]
    },
    {
        code: "foobar4",
        name: { fi: "foo4", sv: "bar4", en: "foobar4" },
        type: EDimensionType.Other,
        values: [
            {
                code: "barfoo1",
                name: { fi: "fyy1", sv: "bör1", en: "fyybör1" },
                isVirtual: false
            },
            {
                code: "barfoo2",
                name: { fi: "fyy2", sv: "bör2", en: "fyybör2" },
                isVirtual: false
            }
        ]
    },
    {
        code: "foobar5",
        name: { fi: "foo5", sv: "bar5", en: "foobar5" },
        type: EDimensionType.Other,
        values: [
            {
                code: "barfoo1",
                name: { fi: "fyy1", sv: "bör1", en: "fyybör1" },
                isVirtual: false
            }
        ]
    }
]

const mockQuery: Query = {
    foobar1: {
        valueFilter: {
            type: FilterType.Item,
            query: 2
        },
        selectable: false,
        virtualValueDefinitions: null
    },
    foobar2: {
        valueFilter: {
            type: FilterType.Item,
            query: 2
        },
        selectable: false,
        virtualValueDefinitions: null
    },
    foobar3: {
        valueFilter: {
            type: FilterType.Item,
            query: 2
        },
        selectable: false,
        virtualValueDefinitions: null
    },
    foobar4: {
        valueFilter: {
            type: FilterType.Item,
            query: 2
        },
        selectable: false,
        virtualValueDefinitions: null
    },
    foobar5: {
        valueFilter: {
            type: FilterType.Item,
            query: 1
        },
        selectable: false,
        virtualValueDefinitions: null
    }
};

const setLanguage = jest.fn();
const language = 'fi';
const setLanguageTab = jest.fn();
const languageTab = 'fi';
const availableUiLanguages = ['fi', 'en', 'sv'];
const uiContentLanguage = 'fi';
const setUiContentLanguage = jest.fn();

describe('Rendering test', () => {
    it('renders correctly', () => {
        const mockSettingsChangedHandler = jest.fn();
        const { asFragment } = render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={mockVisualizationSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion tests', () => {
    it('table settings component should have 3 buttons, up, down and switch', () => {
        const mockSettingsChangedHandler = jest.fn();
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={mockVisualizationSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        expect(screen.getAllByRole('button').length).toBe(7); // up, down, switch and 4 variables === 7
    });

    it('Should call settingschangedhandler when moving row up', async () => {
        const mockSettingsChangedHandler = jest.fn();
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={mockVisualizationSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        fireEvent.click(screen.getByText('foo3 2'));
        fireEvent.click(screen.getAllByText((_content, element) => element.tagName.toLowerCase() === 'button')[0]);
        await waitFor(() => {
            expect(mockSettingsChangedHandler).toBeCalled();
        });
    });

    it('Should call settingschangedhandler when moving col up', async () => {
        const mockSettingsChangedHandler = jest.fn();
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={mockVisualizationSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        fireEvent.click(screen.getByText('foo4 2'));
        fireEvent.click(screen.getAllByText((_content, element) => element.tagName.toLowerCase() === 'button')[0]);
        await waitFor(() => {
            expect(mockSettingsChangedHandler).toBeCalled();
        });
    });

    it('Should call settingschangedhandler when moving row down', async () => {
        const mockSettingsChangedHandler = jest.fn();
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={mockVisualizationSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        fireEvent.click(screen.getByText('foo1 2'));
        fireEvent.click(screen.getAllByText((_content, element) => element.tagName.toLowerCase() === 'button')[1]);
        await waitFor(() => {
            expect(mockSettingsChangedHandler).toBeCalled();
        });
    });

    it('Should call settingschangedhandler when moving col down', async () => {
        const mockSettingsChangedHandler = jest.fn();
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={mockVisualizationSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        fireEvent.click(screen.getByText('foo2 2'));
        fireEvent.click(screen.getAllByText((_content, element) => element.tagName.toLowerCase() === 'button')[1]);
        await waitFor(() => {
            expect(mockSettingsChangedHandler).toBeCalled();
        });
    });

    it('Should call settingschangedhandler when moving row', async () => {
        const mockSettingsChangedHandler = jest.fn();
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={mockVisualizationSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        fireEvent.click(screen.getByText('foo3 2'));
        fireEvent.click(screen.getAllByText((_content, element) => element.tagName.toLowerCase() === 'button')[2]);
        await waitFor(() => {
            expect(mockSettingsChangedHandler).toBeCalled();
        });
    });

    it('Should call settingschangedhandler when moving column', async () => {
        const mockSettingsChangedHandler = jest.fn();
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={mockVisualizationSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        fireEvent.click(screen.getByText('foo4 2'));
        fireEvent.click(screen.getAllByText((_content, element) => element.tagName.toLowerCase() === 'button')[2]);
        await waitFor(() => {
            expect(mockSettingsChangedHandler).toBeCalled();
        });
    });

    it('Shouldnt call settingschangedhandler if nothing selected', async () => {
        const mockSettingsChangedHandler = jest.fn();
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={mockVisualizationSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        fireEvent.click(screen.getAllByText((_content, element) => element.tagName.toLowerCase() === 'button')[0]);
        await waitFor(() => {
            expect(mockSettingsChangedHandler).not.toBeCalled();
        });
    });

    it('When a column variable code is provided in the settings parameter a corresponding list items should be rendered in the column variables list', () => {
        const mockSettingsChangedHandler = jest.fn();
        const toggledMockSettings = { ...mockVisualizationSettings, cutYAxis: true }
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={toggledMockSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        expect(screen.getByText('chartSettings.columnVariables').parentElement).toContainElement(screen.getByText("foo2 2"));
    });

    it('When a row variable code is provided in the settings parameter a corresponding list items should be rendered in the row variables list', () => {
        const mockSettingsChangedHandler = jest.fn();
        const toggledMockSettings = { ...mockVisualizationSettings, cutYAxis: true }
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={toggledMockSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        expect(screen.getByText('chartSettings.rowVariables').parentElement).toContainElement(screen.getByText("foo1 2"));
    });

    it('When a table with single value variables is created, single value variables should not show up in pivot settings', () => {
        const mockSettingsChangedHandler = jest.fn();
        const toggledMockSettings = { ...mockVisualizationSettings, columnVariableCodes: ["foobar2", "foobar5"] }
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={toggledMockSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        expect(screen.getByText('chartSettings.columnVariables').parentElement).toHaveTextContent("foo2 2");
        expect(screen.getByText('chartSettings.columnVariables').parentElement).not.toHaveTextContent("foo5 1");
    });

    it('When a table with a selectable multivalue variable is created, said variable should not show up in pivot settings', () => {
        const mockSettingsChangedHandler = jest.fn();
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={mockVisualizationSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    selectableDimensions={[mockDimensions[1]]}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        expect(screen.getByText('chartSettings.columnVariables').parentElement).not.toHaveTextContent("foo2 2");
    });

    it('When a table with a single value variable with a filter type "from" or "all" is created, said variable should show up in the pivot settings', () => {
        const mockSettingsChangedHandler = jest.fn();
        const toggledMockSettings = { ...mockVisualizationSettings, columnVariableCodes: ["foobar2", "foobar5"] }
        const mockQueryWithPotentiallyGrowingVariable = { ...mockQuery, foobar5: { ...mockQuery.foobar5, valueFilter: { type: FilterType.All } } }
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={mockVisualizationRules}
                    visualizationSettings={toggledMockSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQueryWithPotentiallyGrowingVariable}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        expect(screen.getByText('chartSettings.columnVariables').parentElement).toHaveTextContent("foo2 2");
        expect(screen.getByText('chartSettings.columnVariables').parentElement).toHaveTextContent("foo5 1");
    });

    it('When a table with with a selectable multivalue variable that is set as the multiselectable variable is created, said variable should show up in the pivot settings', () => {
        const mockSettingsChangedHandler = jest.fn();
        const multiselectableMockSettings = { ...mockVisualizationSettings, multiselectableVariableCode: "foobar2" };
        const multiselectableVisualizatioRules = { ...mockVisualizationRules, multiselectVariableAllowed: true };
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <TablePivotSettings
                    dimensions={mockDimensions}
                    visualizationRules={multiselectableVisualizatioRules}
                    visualizationSettings={multiselectableMockSettings}
                    settingsChangedHandler={mockSettingsChangedHandler}
                    query={mockQuery}
                ></TablePivotSettings>
            </UiLanguageContext.Provider>);
        expect(screen.getByText('chartSettings.columnVariables').parentElement).toHaveTextContent("foo2 2");
    });
});