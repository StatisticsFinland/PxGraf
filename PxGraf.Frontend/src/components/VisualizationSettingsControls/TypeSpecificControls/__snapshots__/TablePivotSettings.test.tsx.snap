// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <div
      class="MuiStack-root sc-gEvEer dSbliT css-m69qwo-MuiStack-root"
    >
      <ul
        class="MuiList-root MuiList-padding MuiList-dense MuiList-subheader sc-aXZVg htWJIQ css-p64g7a-MuiList-root"
      >
        <li
          class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-uob957-MuiListSubheader-root"
        >
          chartSettings.rowVariables
        </li>
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
            >
              foo1 2
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
            >
              foo3 2
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </ul>
      <div
        class="MuiStack-root sc-eqUAAy lphtNa css-nen11g-MuiStack-root"
      >
        <button
          aria-label="tableSettings.up"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowUpwardIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m4 12 1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="tableSettings.down"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowDownwardIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="tableSettings.swap"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SwapHorizIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6.99 11 3 15l3.99 4v-3H14v-2H6.99zM21 9l-3.99-4v3H10v2h7.01v3z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <ul
        class="MuiList-root MuiList-padding MuiList-dense MuiList-subheader sc-aXZVg htWJIQ css-p64g7a-MuiList-root"
      >
        <li
          class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-uob957-MuiListSubheader-root"
        >
          chartSettings.columnVariables
        </li>
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
            >
              foo2 2
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
            >
              foo4 2
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
